Let's improve support power timer display.

deniz1a opened the pull request:
Hello. I made a new PR. This is a simpler reimplementation of #7737. It organizes support power countdown timers in teams. Thank you.

deniz1a committed:
Improves support power timer display.
Shows player names, groups timers in teams.

penev92 commented:
Have you given up on the idea of getting the maximum name length in the constructor and using that? Because you have code commented out there.

deniz1a commented:
I tried getting maxPowerNameWidth in constructor but couldn't do it. maxPlayerNameWidth could be in constructor but this way the texts align dynamically. So when a player with a long name doesn't build a support power, it won't be considered in alignment. But to do that this calculates maxPlayerNameWidth every tick.. So if dynamic alignment is not necessary, it could be done using the commented line in constructor.

Which one is better?

Unit158 commented:
Dynamic alignment is better in this case.

Can we get a better name for Line and change it to a struct please?

On 2015-07-02, at 2:14 PM, Deniz <notifications@github.com> wrote:

> I tried getting maxPowerNameWidth in constructor but couldn't do it. maxPlayerNameWidth could be in constructor but this way the texts align dynamically. So when a player with a long name doesn't build a support power, it won't be considered in alignment. But to do that this calculates maxPlayerNameWidth every tick.. So if dynamic alignment is not necessary, it could be done using the commented line in constructor.
> 
> Which one is better?
> 
> ―
> Reply to this email directly or view it on GitHub.
>

deniz1a commented:
How should it be named? I thought line is the most fitting because that's what it represents. Any other suggestions?

Also, isn't class better than struct because it's passed by reference? Although it isn't passed to any method here, just added to a list. So it would probably make no difference? Should I change it to struct?

deniz1a commented:
I could rename Line to Text...

deniz1a commented:
And actually you can't change teams during game. Stance change in diplomacy window doesn't change player's team, only individual stances to other players... So I can move playersByTeam definition to the constructor.

penev92 commented:
This looks like a more complicated version of [this](https://github.com/penev92/OpenRA/blob/ba6b3b8d4315750056db5fa816078289b87c4d53/OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs).
What are the benefits of this one over that?

deniz1a commented:
I don't know, I haven't tested that.

deniz1a commented:
Can Draw() execute before the first Tick()?

penev92 commented:
OK, I'm going to be blunt.

You already had a PR about this that was way more complicated that it needed to be. Because of that a lot of developers spent *a lot* of time on it, including myself. *Then*, at one point, it reached a dead end. Instead of just closing it, I decided to give it a last chance, and spent around four hours of my time to rewrite it in a much simpler way. (Please keep in mind that even before I did so my opinion was that enough time was wasted on the PR by so many people for no gain.) Ultimately the PR was closed because you didn't want to use my suggested solution.

Now you are opening a new PR about the same thing (which is more than likely unfinished feature-wise). The code looks like a slightly more complicated version of the code I suggested for your previous PR, plus a (unnecessary?) `Tick()` method.
You say that you "haven't even tested" my suggested code. (?)

In light of that, I am going to -1 any solution that is more complicated than the one I proposed, unless you can explain how it is better than mine.

deniz1a commented:
Why didn't you open a new pr with your version?

penev92 commented:
Because I don't consider this as something important, and definitely not something that justifies so many people wasting so much time on. And I still don't. So either we resolve this in a reasonable timeframe this time around, or I will close it.
Also because everyone else seems to be OK with the current state of the timers.

deniz1a commented:
Are Tick() and Draw() methods synchronized? What happens when the game is paused? Tick() stops but Draw() continues, right? So it is better to put game state dependent code in Tick() and leave only drawing code in Draw() to minimize calculations.

deniz1a commented:
If you don't consider this important and don't think this feature should be implemented, then why did you waste so much time, as you put it? Why did you write your own version and then not open a pull request?

penev92 commented:
> Instead of just closing it, I decided to give it a last chance

I figured that was the polite thing to do.
Don't get me wrong, I wanted to just close it and stop wasting everyone's time. And I definitely didn't want to spend 2 hours trying to figure out what your code does and 2 more simplifying it.

deniz1a commented:
Ok, so you have the code that implements team grouping? Open a PR and if that's better than this one, it can be reviewed and merged...

deniz1a commented:
I think this PR is finished and needs review. It does two things:
1) Separates SUPPORT_POWER_TIMER widget definition into ingame-player and ingame-observer. This makes it possible to define different bound offsets for observers who don't have support power icons.
2) Organizes support power countdown timers in player teams. It looks like this:
![sptimer-simple](https://cloud.githubusercontent.com/assets/6599106/8504020/1794039e-21d2-11e5-86cd-9421c404b220.png)

Power names and remaining times are aligned dynamically.
This map can be used to quickly test the timer display:
http://resource.openra.net/maps/3506/

Possible changes/improvements that could be done:
* Change ```class Text``` to ```struct``` as suggested by @Unit158.
* Calculate ```maxPlayerWidth``` once in the constructor for all players and not every tick for the players that have a support power.
* Calculate ```maxPowerWidth``` once in the constructor for all buildable support powers in current map or mod, instead of every tick. (I don't know how to do this.)
* Store ```playersByTeam``` in a data structure like Dictionary<> instead of querying it every tick. Or cache ```IOrderedEnumerable```. (I don't know how to do this.)

I can't think of anything else... OK, let's discuss!

deniz1a commented:
Here are Visual Studio profiling results (.vsps files) for bleed and this PR:
https://drive.google.com/folderview?id=0B_dMvsHPqPZkfnJhR1Ryck44Szh4aDhJMjNHd2oySnpsbDNxdjd4UEc1MVhzY0NVcUE0RHM&usp=sharing

abcdefg30 commented:
Looks ok to me. :+1: / :white_check_mark:

pchote commented:
Do we want this feature?  The reception has been luke-warm at best.

deniz1a commented:
Doesn't this improve the timers considerably? Without team grouping and player names, I find the timers to be not that useful.

pchote commented:
That is the open question which nobody has really asked yet.  It adds more details, but it also adds a lot more visual clutter to the screen.  If people don't feel that extra detail is needed, then the additional content could be taken as an overall negative instead of improvement.

Unit158 commented:
I'm going to day that player names are definitely a good thing as our colour validation is simply ass. Teams are a bit overkill but whatever.

On 2015-07-13, at 9:16 AM, Paul Chote <notifications@github.com> wrote:

> That is the open question which nobody has really asked yet. It adds more details, but it also adds a lot more visual clutter to the screen. If people don't feel that extra detail is needed, then the additional content could be taken as an overall negative instead of improvement.
> 
> ―
> Reply to this email directly or view it on GitHub.
>

penev92 commented:
Popular opinion has been that we do not want this.
Also :-1: until you can convince me this really needs `Tick()` *and* `Draw()` methods.

GraionDilach commented:
I'm worried about that this can and will also run out of screen, increasing #7855 just even further.

abcdefg30 commented:
If we remove the `Team` stuff this is kinda just a bigger version of #6282.

deniz1a commented:
I don't think team labels add that much to the length of the text. There are usually two teams... If that's a problem, team labels could be omitted if total number of teams exceeds a set number.

It is better to have separate Tick and Draw methods. It allows for separation of tasks. Tick does game state dependent logic, generates the actual text that will be drawn. Draw is only responsible for drawing on screen. It gets a list of texts and draws them. This has benefits for the following cases:
* When there is lag and Tick frequency is lower than Draw.
* When game is paused. Tick stops, Draw continues
* If in the future, Tick and Draw frequencies are decoupled, Draw could run at much higher rate than Tick. I think currently it runs with the same frequency as Tick (when there's no lag).

The other reason you listed is that this is more complicated than the version you implemented. If that is so, you should open a PR for it so we have a better implementation. It was nice of you to write that code to help me and I got some ideas from it. But what do you want now?

deniz1a commented:
Apart from all that, the more important issue with support power timers is that drawing that text consumes 15% of CPU time according to the profiling results I posted above. If those numbers are correct, there is a problem with text drawing code. Maybe until that is fixed, countdown timers could be shown in a menu widget?

deniz1a commented:
OpenRA.Graphics.SpriteFont.DrawTextWithContrast() consumes more than 14%.

deniz1a commented:
I ran another profiling test and Draw() of SupportPowerTimerWidget uses 18.46% :worried:

deniz1a commented:
I think Tick() and Draw() always run synchronized, so separating logic doesn't improve performance. But still it's better, right?

deniz1a commented:
Widget's Tick() runs with LocalTick, which is separate from world tick, right? But it's not necessary to run the code in this widget's Tick() when world state hasn't changed. So how can I make it run with world tick instead of local tick?

deniz1a commented:
Another issue with this code is that when game is paused, ready timers don't flash. Because colors are calculated in Tick() which is skipped if world tick hasn't changed. But I think this is not important.

deniz1a commented:
Ok I think the game works like this:
There are three different frequencies or timesteps.
* **world.WorldTick** is the game progress. This stops when game is paused and slows down when there's lag. This is synchronized for all players. Runs every 40 ms or 25 times per second when there's no lag.
* **orderManager.LocalFrameNumber** is increased when the Tick() methods of all the classes are called. This is also 25 per second. I think this is not affected by game progress, so constant.
* **Game.RenderFrame** is the frame number of screen drawing. The frequency of this is as high as the computer can render or the frame limit value in settings menu, whichever is lower. This is increased when the Draw() methods of all the classes are called.

Then there's also network tick which has 1/3 the frequency of LocalFrameNumber. That's when packets are sent I think.
The frequency order is: 
RenderFrame > LocalFrameNumber > WorldTick > network tick

Is all this correct?

So, for this PR, we want to generate timer texts every WorldTick and draw them every RenderFrame which is when Draw() is called.

deniz1a commented:
Although I just tested it and the RenderFrame frequency is around 250 fps and is not limited by frame limit setting...

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
+				return;
+
+			lastTick = world.WorldTick;
+			display = new List<Text>();
```
Creating a new list every tick creates work for the garbage collector.

pchote replied:
If you precalculate the power and player names it becomes cheap to calculate the lengths, and you can then drop this memory churning caching completely.

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
+					if (!spiList.Any())
+						continue;
+
+					var playerWidth = font.Measure(player.PlayerName).X;
```
This won't change at runtime, so you can do it once in the constructor and then cache it.

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
+					var showPlayer = true;
+					foreach (var spi in spiList)
+					{
+						var powerWidth = font.Measure(spi.Info.Description).X;
```
Likewise here - you know all the possible support power names before you start, so you can build a dictionary at startup instead of measuring everything every tick.

deniz1a replied:
I can't get the names of all possible support powers... How can I do that?

pchote replied:
`<Map rules>.<ActorInfos>.SelectMany(ai => ai.Traits.Where(t => t is SupportPowerInfo))`

deniz1a replied:
I used this and it threw TargetInvocationException:
```
maxPowerWidth = world.Map.Rules.Actors.Values
				.Select(ai => ai.Traits.GetOrDefault<SupportPowerInfo>())
				.Max(i => font.Measure(i.Description).X);
```

pchote replied:
It will, yes. Your version is completely different to my hint above, and assumes that every actor type has exactly one support power. That assumption fails, so it throws.

deniz1a replied:
Ok, this doesn't crash but the list is empty:
```
maxPowerWidth = world.Map.Rules.Actors.Values
				.SelectMany(ai => ai.Traits.WithInterface<SupportPowerInfo>())
				.Max(i => font.Measure(i.Description).X);
```

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
+
+			playersByTeam = new SortedDictionary<int, List<Player>>();
```
What is the purpose of the `SortedDictionary` here?

deniz1a replied:
So the teams are returned in order. Normal dictionary returns items in undefined order, right?

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
+			foreach (var team in spManagers.Keys.GroupBy(p => world.LobbyInfo.ClientWithIndex(p.ClientIndex).Team))
+			{
+				playersByTeam[team.Key] = new List<Player>();
+				playersByTeam[team.Key].AddRange(team);
```
You can do this in one line (and then drop the braces) with something like `playersByTeam[team.Key] = team.ToList()`.

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
+			this.world = world;
 		}
 
 		public override void Tick()
```
Widget ticking is decoupled from world ticking, so there is no benefit to implementing this in `Tick` compared with inside a `lastTick == world.WorldTick` check inside Render.

deniz1a commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,146 @@
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
```
Should I cache a list of all support power instances that have DisplayTimer true? Instead of sp managers? Does that change during game?

pchote replied:
That won't work, because the instances will change as the game progresses.  What you *can* do is query all the SupportPowerInfos and then store a dictionary of Info to width.

deniz1a replied:
The list of SupportPowerInstance of a player changes during game? How do I get a list of all available powers then?

deniz1a replied:
Does this change during game?
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs#L28

pchote replied:
You can query them from the game rules.

pchote replied:
> Does this change during game?

Yes.  Look at how it is used in that file, and you'll see why/how.

deniz1a replied:
Oh ok thanks.

deniz1a commented:
Refactored the code. Minimized object creation. But I get an NRE when I sell allied tech center.

deniz1a commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,144 @@
+					foreach (var player in team.Value)
+					{
+						var powers = spManagers[player].Powers.Values
+							.Where(i => i.Info.DisplayTimer && i.Instances.Any() && !i.Disabled).ToArray();
```
Does this change while it is still being queried and cause the NRE?

Unit158 replied:
no. the support power will be destroyed before it is removed from the list.

On 2015-07-15, at 10:35 PM, Deniz <notifications@github.com> wrote:

> In OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
> 
> >  			{
> > -				var font = Game.Renderer.Fonts[Font];
> > -				font.DrawTextWithContrast(t.First, new float2(Bounds.Location) + new float2(0, y), t.Second, Color.Black, 1);
> > -				y += (font.Measure(t.First).Y + 5) * (int)Order;
> > +				lastTick = world.WorldTick;
> > +				display.Clear();
> > +				alternateColor = Game.LocalTick % 50 > 25;
> > +				foreach (var team in playersByTeam)
> > +				{
> > +					needTeamLabel = true;
> > +					foreach (var player in team.Value)
> > +					{
> > +						var powers = spManagers[player].Powers.Values
> > +							.Where(i => i.Info.DisplayTimer && i.Instances.Any() && !i.Disabled).ToArray();
> Does this change while it is still being queried and cause the NRE?
> 
> ―
> Reply to this email directly or view it on GitHub.
>

deniz1a commented:
Fixed NRE, added comments.

deniz1a commented:
Looks finished to me, anything else that can be done?

GraionDilach commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,148 @@
-		public readonly string Format = "{0}: {1}";
-		public readonly TimerOrder Order = TimerOrder.Descending;
+		[Translate] public readonly string NoTeamText = "No Team";
+		[Translate] public readonly string TeamText = "Team {0}";
```
Exposing a direct C# format string to YAML is a bad idea. You sure there isn't a better way?

pchote replied:
Unfortunately not, but that's a discussion for our wider translation efforts and is way out of scope of this pr.

GraionDilach replied:
Alright then.

GraionDilach commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,148 @@
-				y += (font.Measure(t.First).Y + 5) * (int)Order;
+				lastTick = world.WorldTick;
+				display.Clear();
+				alternateColor = Game.LocalTick % 50 > 25;
```
I see you carried this over but why not using 60 and 30 as magic values instead? Your changes turn ready texts from a balanced 50:50 blinking to one which prefers the player's color.

deniz1a replied:
I think 50 and 25 were used because local tick frequency is 25 per second, if there is no lag. So this would ideally show player color and flashing color for one second each.

GraionDilach replied:
"Ideally", but then you only check every "0.4 seconds" leading to 1.2 for one color and 0.8 for the other.

deniz1a replied:
I changed that, now text flashing is done in draw part of the method. So it will flash even when the game is paused.

GraionDilach commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,148 @@
-			foreach (var t in texts)
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick - lastTick > 9)
```
This should be replaced with a (world.WorldTick % 10 == 0) check and lastTick should be removed then. The kind of optimization you're doing here should be done by the VM if needed.

deniz1a replied:
I tried that and it seems to skip some ticks.. The countdown seconds look irregular, I don't know why?

GraionDilach replied:
Ah, I see, it must be a race condition when Tick is faster than rendering - in that scenario your code might be a better option indeed.

GraionDilach replied:
Another option worths a try is to use Game.LocalTick here to prevent that race condition.

deniz1a committed:
Improves ready text flashing.

deniz1a commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,150 @@
+			public string Player;
+			public string Power;
+			public string RemainingTime;
+			public bool Ready;
```
Instead of having a separate Ready variable here, I could compare RemainingTime to zero time. I don't know which one is better.

deniz1a commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -11,55 +11,150 @@
-			foreach (var t in texts)
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick % 10 == 0)
```
The game cannot skip a world tick, right? Because that could cause this to not update often enough.

GraionDilach replied:
>it must be a race condition when Tick is faster than rendering

Between your draws, the Ticks already pass. Either that, either the UI is separated from the world in this case.

Considering that the ready-text-blinking code uses Game.LocalTick instead, I'd assume that's much reliable than this. As I said it already.

deniz1a replied:
But that's different. Ready text blinking and countdown timer flashing are only visual indications. They don't depend on world state.

pchote commented:
This has been modified since the original +1, so removing the needs +2 label (needs a full re-review).

obrakmann commented:
:-1: I don't want this.

In fact, I'd rather see the support power timers reduced to your own (ie. just the icons), and if need be your allies. Why should my enemies know when my spy satellite launches?

GraionDilach commented:
SpySat is a bad use-case. Knowing your enemy's Nuke/IonCannon is important. Not sure if we have a tag for hiding the countdown.

deniz1a commented:
Actually I also think the countdown timers should be shown only for allies. But with this PR I was trying to improve the current display without changing any behaviour. Should I close this PR and open one that reduces timers to only those belonging to allies? Or this PR could still be useful for observers. So I could add that feature to this one.

deniz1a committed:
Removes timers that don't belong to allies.
Also removes player's own timers.

deniz1a commented:
Ok I removed timers that belong to players who are not allied with current player. Also removed player's own timers, they are already shown in icons.

abcdefg30 commented:
Sorry, but :-1: for only showing timers of allied players.

deniz1a commented:
It's more logical and also improves gameplay I think.

obrakmann commented:
OK, it seems we can't come to an agreement over this, so it's probably best to stay at the status quo.

deniz1a commented:
I reverted the commit that hides enemy timers. I think there is an agreement that this would improve the game, the only part with disagreement was hiding enemy timers and I removed it.

deniz1a commented:
I'm keeping my branch upto date with bleed, in case an agreement is reached that this is better than current timers. Code is here:
https://github.com/deniz1a/OpenRA/tree/countdown-simple

GraionDilach commented:
Add something like a DisplayTimerToEnemies, set it false on GPS, true on AtomBomb and I think this is good then.

deniz1a commented:
Should I open a PR with those changes? Why should atom bomb timer be shown to enemies but GPS not?

GraionDilach commented:
Honestly I don't even understand why the GPS timer is even shown to the player, but meh. I'd consider the GPS as a support power (equal of Parabombs, Paratroopers or the Sonar Pulse) which does not majorly affect the game that much, compared to Chronosphere, Iron Curtain or the A-Bomb.

deniz1a commented:
I think it affects the gameplay but those timers should be shown only to allies. I'll add DisplayTimerToEnemies and maybe open another PR.

deniz1a commented:
I added ````DisplayTimerToEnemies```` to ````SupportPower````, so now it's configurable in yaml.

I think if there should be countdown timers in game, my countdown-simple branch is the way to do it. Even though the code is a lot more complex than the current implementation, there is no simpler way of doing it right. And despite the added logic, it is very efficient. It might even be faster to run than the current code, aside from any performance issues in font drawing.

In conclusion, I present to you this marvel of software engineering which improves the player experience significantly. The open question now is, if countdown timers are an important enough feature to justify maintaining this code. I think they are.

pchote commented:
Please do not open another pr.  For the reasons that we have been over (and over and over...) across the last two PRs and the IRC channel this is not as straightforward as you claim. I think I can speak for everyone on the reviewing side by saying that we are not willing to reshash these arguments again.

In conclusion, no.  Lets not improve the support power display.  The timers are not important enough to justify the amount of hair that you are making us pull out in frustration over this.
